content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx

```diff
@@ -0,0 +1,401 @@
+}
+
+async function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithFullName = computeFullName(user)
+}
+```
+
+In the example above, you've defined a `User` generic that extends the `FirstLastName` type. This means whatever you pass into `computeFullName` must contain `firstName` and `lastName` keys.
+
+You've also defined a `WithFullName<User>` return type, which takes whatever `User` is and tacks on a `fullName` string attribute.
+
+With your function, any object that contains `firstName` and `lastName` keys can compute a `fullName`. Pretty neat, right?
+
+## Excluding **Sensitive** Fields
+
+Very often, you don't want to include the **whole response** back from the database. In the signup function above, we've included the `password` in the response. Let's exclude this sensitive piece of data from the response.
+
+### Exclude Password in Javascript
+
+Once again, we'll start with Javascript so you can first focus on the logic. You first want to create an `exclude` helper function that takes a user and returns the user without certain fields:
+
+```tsx file=index.ts
+// Remove keys from user
+function exclude(user, ...keys) {
+  for (let key of keys) {
+    delete user[key]
+  }
+  return user
+}
+
+function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+### Exclude Password in Typescript
+
+Now let's make this `exclude` function type-safe. You first want to make sure that the keys you provide exist on the user and then you want to return a user that doesn't contain those keys.
+
+Here's what that would look like in Typescript:
+
+```tsx file=index.ts highlight=1-4;normal
+function exclude<User, Key extends keyof User>(
+  user: User,
+  ...keys: Key[]
+): Omit<User, Key> {
+  for (let key of keys) {
+    delete user[key]
+  }
+  return user
+}
+
+function main() {
+  const user = await prisma.user.findUnique({ where: 1 })
+  const userWithoutPassword = exclude(user, 'password')
+}
+```
+
+In this example, we've provided two generics `User` and `Key`. The `Key` generic is defined as the keys of a `User` (e.g. email, password, firstName, etc.). These generics flow through the logic, returning a User that omits the list of keys provided.
+
+## Organizing the Prisma Client **into Models**
+
+In the examples above, you've created helper functions that weave the type-safety deeper through your application. Creating helper functions works well when your application is small, but as your application grows, you may find an inconsistent use of these helper functions strewn across your codebase.
+
+One technique for consistently applying and organizing these helper functions is to group them in a class.
+
+### Wrapping the Model in a Class
+
+The most obvious way to wrap the Prisma Client is by wrapping it in a `Model` class:
+
+```tsx highlight=4,23,28-30;normal
+import { PrismaClient } from '@prisma/client'
+
+class Model {
```

Comment:
Maybe rename "Model" to "UserService"? I would expect a user model to be a custom representation of a user (with name, email etc as properties) but not a wrapper around prisma client.

Reply:
Terminology is very tricky here. From Prisma's perspective a model represents a table or collection you can query on and what you get back is a model result. 

I like your definition of a model a bit more, which matches my understanding that a model is an item in a collection that has properties and methods. 

I find UserService overly generic, but since I don't have any better ideas, I'm going to cheat a bit and pluralize the function to avoid definition altogether.

```ts
async function main() {
  const prisma = new PrismaClient()
  const users = Users(prisma.user)
  const user = await users.signup(/* { ... } */)
  const numUsers = await users.count()
  console.log(user, numUsers)
}
```

Thanks for this comment :-)